Custom urls routing

wachterjohannes opened the pull request:
| Q | A
| --- | ---
| Bug fix? | no yes
| New feature? | no yes
| BC breaks? | no yes
| Deprecations? | no yes <!-- if yes add them to the UPGRADE.md file -->
| Fixed tickets | fixes # <!-- add issue number here e.g.: #5730 -->
| Related issues/PRs | # <!-- add issue or PR number here e.g.: #5730 -->
| License | MIT
| Documentation PR | sulu/sulu-docs# <!-- add docs PR number here e.g.: sulu/sulu-docs#615 -->

#### What's in this PR?

<!-- Explain the contents of the PR. -->

#### Why?

<!-- Which problem does the PR fix? (add some context and maybe link to an issue here) -->

#### Example Usage

<!--
```php
// If you added new features, show examples of how to use them here

$foo = new Foo();

// Now we can do
$foo->doSomething();
```
-->

#### To Do

- [ ] Create a documentation PR
- [ ] Add breaking changes to UPGRADE.md

<!--

Dear Contributors,

Thank you for contributing to the Sulu ecosystem!

We appreciate your effort to improve our project.  
If you need assistance or have questions about your pull request, our team is here to help.  
Please join our Slack channel for support: https://sulu.io/services/support#chat.

Best Regards, 
The Sulu Core Team

-->

mamazu commented on packages/custom-url/src/Application/Routing/CustomUrlRouteCollectionLoader.php:
```diff
@@ -0,0 +1,196 @@
+        $pathInfo = $request->getPathInfo();
+
+        // Remove format extension if present (e.g., .html, .json)
+        $position = \strrpos($pathInfo, '.');
```
This would also break `/test.hallo/foo`

mamazu commented on packages/custom-url/src/Domain/Model/CustomUrlRoute.php:
```diff
@@ -48,4 +52,24 @@ public function getCustomUrl(): CustomUrlInterface
+        $this->history = $history;
+    }
+
+    public function getTargetRoute(): ?CustomUrlRouteInterface
+    {
+        return $this->targetRoute;
+    }
+
+    public function setTargetRoute(?CustomUrlRouteInterface $targetRoute): void
+    {
+        $this->targetRoute = $targetRoute;
+    }
```
Isn't this a property of the custom url already?

wachterjohannes replied:
no we have to know which is the followup route to have the chain which is the newest route

mamazu commented on packages/custom-url/src/Infrastructure/Doctrine/EventListener/CustomUrlRouteHistoryListener.php:
```diff
@@ -0,0 +1,137 @@
+     */
+    private array $pendingHistoryUpdates = [];
+
+    public function preUpdate(PreUpdateEventArgs $args): void
```
This only needs to happen on CustomUrl update? Can't we use the domain event for that? Ohterwise this will run for any entity update which seems a little excessive.

wachterjohannes replied:
i am currently refactoring this :)

alexander-schranz replied:
the listener does nothing if no customurl is changed and it should do nothing if no url was changed.

You want todo the history url creation in the same transaction and not in an own. Also you want to trigger history url creation also when somebody directly modify the entity not using the our provided messages. Its to critical todo it on the domain event.

alexander-schranz commented on packages/custom-url/src/Infrastructure/Doctrine/EventListener/CustomUrlRouteHistoryListener.php:
```diff
@@ -0,0 +1,137 @@
+ * Manages history routes when custom URL paths change.
+ * Marks old routes as history and links them to new routes.
+ */
+class CustomUrlRouteHistoryListener
```
```suggestion
/**
 * @internal No BC promises are given for this class. It may be changed or removed at any time.
 *
 * Manages history routes when custom URL paths change.
 * Marks old routes as history and links them to new routes.
 */
class CustomUrlRouteHistoryListener
```

wachterjohannes replied:
have refactored this into the modify listener as this does not make sense to handle that in a doctrine listener

alexander-schranz replied:
it does make sense if custom url get modified directly to still create route histories.

alexander-schranz commented on packages/custom-url/src/Infrastructure/Doctrine/EventListener/CustomUrlRouteHistoryListener.php:
```diff
@@ -0,0 +1,137 @@
+    /**
+     * @var array<string, array{oldPath: string, newPath: string, customUrl: CustomUrl}>
+     */
+    private array $pendingHistoryUpdates = [];
```
service requires reset interface

alexander-schranz replied:
requires also onclear listener

wachterjohannes committed:
Implement complete custom-URL routing based on Sulu 2.6 architecture

This is a complete rewrite of custom-URL routing that replicates all
functionality from Sulu 2.6 while adapting to Sulu 3.0's architecture.

Complete feature implementation:
- Route matching by full URL path (host + path)
- History route management with automatic 301 redirects
- URL normalization (trailing slash and .html redirects)
- Redirect handling with WebspaceManager URL building
- Content display with ContentAggregator and template resolution
- Full SEO support: redirect, canonical, noFollow, noIndex

Architecture changes from Sulu 2.6 to 3.0:
- PHPCR RouteDocument → Doctrine CustomUrlRoute entity
- PageBridge/Structure → ContentAggregator/DimensionContent
- StructureManager → MetadataProviderRegistry
- Route enhancer chain → Integrated priority-based handling
- CustomUrlRequestProcessor → Direct routing in loader

Priority-based processing (matches 2.6 enhancer chain):
1. Priority 100: Trailing slash normalization
2. Priority 90: Trailing .html normalization
3. Priority 80: Redirect handling (isRedirect or no target)
4. Priority 70: SEO properties (canonical, noFollow, noIndex)
5. Priority 60-40: Content display with template resolution

Technical implementation:
- WebspaceManager integration for proper URL building
- EntityManager for Page entity loading by UUID
- ContentAggregator for live dimension content resolution
- MetadataProviderRegistry for template metadata (controller/view)
- History routes with self-referencing targetRoute
- Format extension and query string preservation
- Published state validation

Database schema:
- Added history field to CustomUrlRoute (boolean, default false)
- Added targetRoute self-reference for history chains
- Indexed on path and history for efficient lookups
- Foreign key with ON DELETE SET NULL for broken chains

Files modified:
- CustomUrlRoute.php: Added history and targetRoute support
- CustomUrlRoute.orm.xml: Added fields and indexes
- CustomUrlRepository: Added findRouteByPath and findHistoryRoutes
- CustomUrlRouteCollectionLoader: Complete rewrite with full functionality
- CustomUrlRouteHistoryListener: Automatic history management
- SuluCustomUrlBundle: Updated service registration
- UPGRADE-3.x.md: SQL migration documentation

alexander-schranz commented on packages/custom-url/config/doctrine/CustomUrl/CustomUrlRoute.orm.xml:
```diff
@@ -9,13 +9,23 @@
         <indexes>
-            <index name="routes" columns="path" />
+            <index name="idx_route_path" columns="path" />
+            <index name="idx_route_history" columns="history" />
```
Index name need to be unique over the DB and shoould so have table name as a prefix.

```suggestion
            <index name="custom_url_route_path_idx" fields="path" />
            <index name="custom_url_route_history_idx" fields="history" />
```

similar to: https://github.com/sulu/sulu/blob/4ef044f0223ee0c814f295f87fb6e7ff5076a80c/packages/route/config/doctrine/Route/Route.orm.xml#L14

alexander-schranz commented on packages/custom-url/config/doctrine/CustomUrl/CustomUrlRoute.orm.xml:
```diff
@@ -9,13 +9,23 @@
          </many-to-one>
 
+        <many-to-one field="targetRoute" target-entity="Sulu\CustomUrl\Domain\Model\CustomUrlRouteInterface">
+            <join-column name="target_route_uuid" referenced-column-name="uuid" nullable="true" on-delete="set null" />
```
why not `CASCADE` here if the target is removed?

wachterjohannes replied:
absolutly right :)

wachterjohannes committed:
fix code-review
